File: web/src/engine/Palette.ts
import type { RGBA8 } from './types'

export interface PaletteColour {
  rgba: RGBA8
  number: number  // 1-based display label
}

export class Palette {
  colours: PaletteColour[]
  readonly originalColours: PaletteColour[]

  constructor(colours: PaletteColour[]) {
    this.colours = colours.map(c => ({ ...c, rgba: { ...c.rgba } }))
    this.originalColours = colours.map(c => ({ ...c, rgba: { ...c.rgba } }))
  }

  replaceColour(index: number, rgba: RGBA8): void {
    if (index >= 0 && index < this.colours.length) {
      this.colours[index] = { ...this.colours[index], rgba: { ...rgba } }
    }
  }

  reset(): void {
    this.colours = this.originalColours.map(c => ({ ...c, rgba: { ...c.rgba } }))
  }

  cssString(index: number): string {
    const c = this.colours[index]
    if (!c) return '#f5f5f5'
    return `rgb(${c.rgba.r},${c.rgba.g},${c.rgba.b})`
  }

  toJSON(): PaletteColour[] {
    return this.colours.map(c => ({ ...c, rgba: { ...c.rgba } }))
  }

  static fromJSON(data: PaletteColour[]): Palette {
    return new Palette(data)
  }
}

File: TappyColour/Engine/Palette.swift
import Foundation
import SwiftUI

// MARK: - PaletteColour
struct PaletteColour: Equatable, Codable {
    var rgba: RGBA8
    let number: Int

    init(rgba: RGBA8, number: Int) {
        self.rgba = rgba
        self.number = number
    }
}

// MARK: - Palette
struct Palette: Equatable, Codable {
    var colours: [PaletteColour]
    let originalColours: [PaletteColour]

    init(colours: [PaletteColour], originalColours: [PaletteColour]) {
        self.colours = colours
        self.originalColours = originalColours
    }

    // MARK: Colour Manipulation
    mutating func replaceColour(at index: Int, with rgba: RGBA8) {
        guard index >= 0, index < colours.count else { return }
        colours[index].rgba = rgba
    }

    mutating func reset() {
        colours = originalColours
    }

    // MARK: SwiftUI Integration
    func swiftUIColor(at index: Int) -> Color {
        guard index >= 0, index < colours.count else {
            return Color.clear
        }
        let colour = colours[index]
        let red = Double(colour.rgba.r) / 255.0
        let green = Double(colour.rgba.g) / 255.0
        let blue = Double(colour.rgba.b) / 255.0
        let alpha = Double(colour.rgba.a) / 255.0
        return Color(red: red, green: green, blue: blue, opacity: alpha)
    }
}
